tokio/src/macros/select.rs

```diff
@@ -0,0 +1,788 @@
+/// Wait on multiple concurrent branches, returning when the **first** branch
+/// completes, cancelling the remaining branches.
+///
+/// The `select` macro accepts one or more branches with the following pattern:
+///
+/// ```text
+/// <pattern> = <async expression> (, if <condition>)? => <handler>,
+/// ```
+///
+/// Additionally, the `select!` macro may include a single, optional `else`
+/// branch, which evaluates if none of the other branches match their patterns:
+///
+/// ```text
+/// else <expression>
+/// ```
+///
+/// The macro aggregates all `<async expression>` expressions and runs them
+/// concurrently on the **current** task. Once the **first** expression
+/// completes with a value that matches its `<pattern>`, the `select!` macro
+/// returns the result of evaluating the completed branch's `<handler>`
+/// expression.
+///
+/// Additionally, each branch may include an optional `if` condition. This
+/// condition is evaluated **before** the <async expression>. If the condition
+/// returns `false`, the branch is entirely disabled. This capability is useful
+/// when using `select!` within a loop.
+///
+/// The complete lifecycle of a `select!` expression is as follows:
+///
+/// 1. Aggregate the `<async expression>`s from each branch.
+/// 2. Evaluate all provded `<condition>` expressions. If the condition returns
+///    `false`, disable the branch.
+/// 3. Concurrently await on the results for all remaining `<async expression>s.
+/// 4. Once an `<async expression>` returns a value, attempt to apply the value
+///    to the provided `<pattern>`, if the pattern matches, evaluate `<handler>`
+///    and return. If the pattern **does not** match, disable the current branch
+///    and continue from step 3.
+/// 5. If **all** branches are disabled, evaluate the `else` expression. If none
+///    if provided, panic.
+///
+/// # Notes
+///
+/// By running all async expressions on the current task, the expressions are
+/// able to run **concurrently** but not in **parallel**. This means all
+/// expressions are run on the same thread and if one branch blocks the thread,
+/// all other expressions will be unable to continue.
+///
```

Comment:
I don’t follow. That example should work?

Reply:
So the issue is that the "completion" might be lost, since the state of `is_elapsed` is updated even if the delay is never polled, like shown here:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bf8db1e28b45346f9eb40813988fed24

That branch of the `select!` would be racy, and you might sporadically see it complete depending on what else is going on. It's the same reasoning for why the `FusedFuture` impl for `Delay` was reverted in #1652 (after a conversation on Discord).

This can be a rather surprising gotcha, depending on how the future is implemented.

Reply:
I added the example & notes.